app/src/main/java/com/abhijitvalluri/android/fitnotifications/services/NLService.java

```diff
@@ -214,185 +204,74 @@ public void onNotificationPosted(final StatusBarNotification sbn) {
-                if (temp.contains(filterText)) { // This notification should not be sent
-                    return;
-                }
-            }
-
-            if (mTransliterateNotif) {
-                notificationText = Transliterator.getInstance("Any-Latin")
-                        .transform(extras.getCharSequence(Notification.EXTRA_TEXT).toString());
-            } else {
-                notificationText = extras.getCharSequence(Notification.EXTRA_TEXT).toString();
-            }
-        } catch (NullPointerException e) {
-            notificationText = "";
-        }
-
-        try {
-            String temp = extras.getCharSequence(Notification.EXTRA_BIG_TEXT).toString();
-            if (!filterText.isEmpty()) {
-                if (temp.contains(filterText)) { // This notification should not be sent
-                    return;
-                }
-            }
-
-            if (mTransliterateNotif) {
-                notificationBigText = Transliterator.getInstance("Any-Latin")
-                        .transform(extras.getCharSequence(Notification.EXTRA_BIG_TEXT).toString()); // TODO: Apparently need minimum API 21 to use EXTRA_BIG_TEXT
-            } else {
-                notificationBigText = extras.getCharSequence(Notification.EXTRA_BIG_TEXT).toString();
-            }
-        } catch (NullPointerException e) {
-            notificationBigText = "";
-        }
+        CharSequence notificationTitle = extras.getCharSequence(Notification.EXTRA_TITLE);
+        String notificationText = buildNotificationText(extras, appPackageName, discardEmptyNotifications);
 
-        try {
-            String temp = extras.getCharSequence(Notification.EXTRA_TITLE).toString();
-            if (!filterText.isEmpty()) {
-                if (temp.contains(filterText)) { // This notification should not be sent
-                    return;
-                }
-            }
-        } catch (NullPointerException e) {
-            // Do nothing and just continue
-        }
-
-        if (discardEmptyNotifications && notificationText.trim().isEmpty() && notificationBigText.trim().isEmpty()) {
+        // notificationText can be null only when discardEmptyNotifications is enabled
+        if (notificationText == null || anyMatchesFilter(filterText, notificationTitle, notificationText)) {
             return;
         }
 
-        if (notificationBigText.length() > 0 && notificationBigText.startsWith(notificationText)) {
-            notificationBigText = notificationBigText.substring(notificationText.length());
-        }
-
-        StringBuilder sb = new StringBuilder();
-        if (mDisplayAppName) {
-            sb.append("[")
-              .append(mAppSelectionsStore.getAppName(appPackageName))
-              .append("] ");
+        String prevNotificationText = mNotificationStringMap.put(appPackageName, notificationText);
+        // TODO: add more specific checks to avoid blocking legitimate identical notifications
```

Comment:
The reason there is a duplicated notification is because if someone makes use of the new "reply" feature from the notification on newer Android builds, then doing so causes the same notification to be issued again. There may be a better way to check for duplicates, but for now this does the job.